File: ABC/ABC192/E.py
import sys
from heapq import heappop, heappush

sys.setrecursionlimit(10 ** 7)
rl = sys.stdin.readline


def solve():
    N, M, X, Y = map(int, rl().split())
    X, Y = X - 1, Y - 1
    graph = [[] for _ in range(N)]
    T, K = [0] * M, [0] * M
    for i in range(M):
        a, b, t, k = map(int, rl().split())
        a, b = a - 1, b - 1
        graph[a].append((b, i))
        graph[b].append((a, i))
        T[i] = t
        K[i] = k
    
    dist = [10 ** 18] * N
    dist[X] = 0
    que = [(0, X)]
    while que:
        time, cur = heappop(que)
        if cur == Y:
            print(time)
            return
        if dist[cur] < time:
            continue
        for to, idx in graph[cur]:
            wait = K[idx] * (time // K[idx] + 1) - time
            if wait == K[idx]:
                wait = 0
            next_t = time + wait + T[idx]
            if dist[to] <= next_t:
                continue
            dist[to] = next_t
            heappush(que, (next_t, to))
    print(-1)


if __name__ == '__main__':
    solve()

File: ABC/ABC192/E.nim
import algorithm, bitops, deques, heapqueue, math, sets, sequtils, strutils, sugar, tables

proc input*(): string = stdin.readLine
proc inputInt*(): int = stdin.readLine.parseInt()
proc inputInts*(): seq[int] = stdin.readLine.split.map(parseInt)
proc chmax*[T: SomeNumber](num0: var T, num1: T) = num0 = max(num0, num1)
proc chmin*[T: SomeNumber](num0: var T, num1: T) = num0 = min(num0, num1)
proc `%=`*[T: SomeInteger](num0: var T, num1: T) = num0 = floorMod(num0, num1)

var
  graph: seq[seq[(int, int)]]
  T, K, dist: seq[int]
  que: HeapQueue[(int, int)]

proc solve() =
  var N, M, X, Y, a, b, t, k: int
  (N, M, X, Y) = inputInts()
  dec X; dec Y
  graph = newSeqWith(N, newSeq[(int, int)]())
  (T, K) = (newSeq[int](M), newSeq[int](M))
  for i in 0..<M:
    (a, b, t, k) = inputInts()
    dec a; dec b
    graph[a].add((b, i))
    graph[b].add((a, i))
    (T[i], K[i]) = (t, k)
  
  dist = newSeqWith(N, 10^18)
  dist[X] = 0
  que = initHeapQueue[(int, int)]()
  que.push((0, X))
  var time, cur, wait: int
  while que.len != 0:
    (time, cur) = que.pop()
    if cur == Y:
      echo time
      return
    if dist[cur] < time:
      continue
    for (to, idx) in graph[cur]:
      wait = K[idx] * (time div K[idx] + 1) - time
      if wait == K[idx]:
        wait = 0
      let next_t = time + wait + T[idx]
      if dist[to] <= next_t:
        continue
      dist[to] = next_t
      que.push((next_t, to))
  echo -1

when is_main_module:
  solve()
